src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,573 @@
+#include "logical_query_plan/aggregate_node.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "statistics/table_statistics.hpp"
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+namespace {
+
+/**
+ * Info about a predicate that needs to be pulled up into a join predicate.
+ */
+struct PredicateInfo {
+  /**
+   * Comparison operand from the left sub-tree.
+   */
+  std::shared_ptr<AbstractExpression> left_operand;
+
+  /**
+   * Comparison operand from the right sub-tree (the previous subquery).
+   */
+  std::shared_ptr<AbstractExpression> right_operand;
+
+  PredicateCondition condition;
+};
+
+/**
+ * Collected information about the predicates that need to be pulled up and other nodes that need to be adjusted.
+ */
+struct PredicatePullUpInfo {
+  /**
+   * Predicates to pull up.
+   *
+   * Ordered by depth in the sub-tree, from top to bottom.
+   */
+  std::vector<PredicateInfo> predicates;
+
+  /**
+   * Nodes that contain the predicates in the subqueries LQP.
+   */
+  std::set<std::shared_ptr<PredicateNode>> predicate_nodes;
+
+  /**
+   * Projection nodes in the sub-tree that need to be removed to make the predicate pull-up safe.
+   */
+  std::vector<std::shared_ptr<ProjectionNode>> projection_nodes;
+
+  /**
+   * Alias nodes in the sub-tree that need to be extended to make the predicate pull-up safe.
+   *
+   * The nodes are ordered top to bottom.
+   */
+  std::vector<std::shared_ptr<AliasNode>> alias_nodes;
+
+  /**
+   * Aggregates that need to be patched allow the predicates to be pulled up.
+   *
+   * Ordered by depth in the sub-tree, from top to bottom. The number notes the number of predicates above this
+   * aggregate in the LQP.
+   */
+  std::vector<std::pair<std::shared_ptr<AggregateNode>, size_t>> aggregate_nodes;
+};
+
+/**
+ * Checks whether a predicate node should be turned into a join predicate.
+ */
+std::optional<PredicateInfo> should_become_join_predicate(
+    const std::shared_ptr<PredicateNode>& predicate_node,
+    const std::map<ParameterID, std::shared_ptr<AbstractExpression>>& parameter_mapping, bool is_below_aggregate) {
+  // Check for the type of expression first. Note that we are not concerned with predicates of other forms using
+  // correlated parameters here. We check for parameter usages that prevent optimization later in
+  // contains_unoptimizable_correlated_parameter_usages.
+  if (predicate_node->predicate()->type != ExpressionType::Predicate) {
+    return std::nullopt;
+  }
+
+  const auto& predicate_expression = std::static_pointer_cast<AbstractPredicateExpression>(predicate_node->predicate());
+
+  // Joins only support these six binary predicates. We rely on PredicateSplitUpRule having split up ANDed chains of
+  // such predicates previously, so that we can process them separately.
+  auto cond_type = predicate_expression->predicate_condition;
```

Comment:
I replaced it by a switch. Maybe this could be factored into a helper method, since `OperatorJoinPredicate::from_expression` does the exact same check.